### PR TITLE
[feature](hive)Support hive tables after alter type.

### PR DESCRIPTION
**Pull request by hubgeter**
## Proposed changes

Issue Number: close #xxx
1.Reconstruct the logic of decode to read parquet. The parquet  reader first reads the data according to the parquet physical type, and then performs a type conversion.

2.Support hive alter table.

|    to\from     |  int     |smallint|tinyint|bigint|float|double|boolean|string|char|varchar|date|timestamp|decimal|
|---------|----------|--------|-------|------|-----|----------------|-------|------|----|-------|----|---------|-------|
|int      | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;      |
|smallint | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;
|tinyint  | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;
|float    | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;
|bigint   | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;
|double   | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;
|boolean  |  ||||||&#10004;
|string   | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|
|char   | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|
|varchar   | &#10004; |&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|&#10004;|
|date     |          | |||||||||&#10004;|
|timestamp|          |||||||||||&#10004;|
|decimal  |&#10004;|&#10004;|&#10004;|&#10004;|||||||||&#10004;|

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hubgeter** [comment]
run buildall

**Comment by wm1581066** [comment]
schema change

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run p0

**Comment by hubgeter** [comment]
run p0

File: be/src/vec/exec/format/parquet/parquet_column_convert.h [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+    auto str_col = static_cast<ColumnString*>(dst_col);
+    for (int i = 0; i < rows; i++) {
+        std::string value = std::to_string(src_data[i]);
+        str_col->insert_data(value.data(), value.size());
+    }
+    return Status::OK();
+}
+
+template <bool is_nullable>
+struct int128totimestamp : public ColumnConvert {
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/parquet_column_convert.h:265]: `Int128ToTimestamp`, use upper case.

File: be/src/vec/exec/format/parquet/parquet_column_convert.h [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+    void convert_null(const IColumn** src_col, IColumn** dst_col) const {
+        size_t rows = (*src_col)->size();
+        if ((*src_col)->is_nullable()) {
+            auto src_nullable_column = static_cast<const ColumnNullable*>(*src_col);
+            auto dst_nullable_column = static_cast<ColumnNullable*>(*dst_col);
+            auto& src_null_data = src_nullable_column->get_null_map_column().get_data();
+            dst_nullable_column->get_null_map_column().resize(_convert_params->start_idx + rows);
+            auto& dst_null_data = dst_nullable_column->get_null_map_column().get_data();
+            for (auto j = 0; j < rows; j++) {
+                dst_null_data[_convert_params->start_idx + j] = src_null_data[j];
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/parquet_column_convert.h:208]: use `memcpy`

File: be/src/vec/exec/format/parquet/parquet_column_convert.h [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+Status convert_data_type_from_parquet(tparquet::Type::type parquet_type, PrimitiveType,
+                                      vectorized::DataTypePtr& ans_data_type, DataTypePtr& src_type,
+                                      bool* need_convert);
+
+struct ColumnConvert {
+    virtual Status convert(const IColumn* src_col, IColumn* dst_col) { return Status::OK(); }
+
+    virtual ~ColumnConvert() = default;
+
+    void convert_null(const IColumn** src_col, IColumn** dst_col) const {
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/parquet_column_convert.h:199]: Why use double `**`(pointer of pointer)? We'd better use `IColumn& col` as the parameter, or `MutableColumnPtr& col` to stand the column to be altered.

File: be/src/vec/exec/format/parquet/parquet_column_convert.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+// under the License.
+
+#include "vec/exec/format/parquet/parquet_column_convert.h"
+
+#include <cctz/time_zone.h>
+
+namespace doris::vectorized {
+namespace ParquetConvert {
+const cctz::time_zone ConvertParams::utc0 = cctz::utc_time_zone();
+
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/parquet_column_convert.cpp:25]: Add comment for `show_type `, `ans_data_type `  and `src_type `

File: be/src/vec/exec/format/parquet/parquet_column_convert.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        break;
+    case tparquet::Type::type::DOUBLE:
+        ans_data_type = std::make_shared<DataTypeFloat64>();
+        break;
+    case tparquet::Type::type::BYTE_ARRAY:
+    case tparquet::Type::type::FIXED_LEN_BYTE_ARRAY:
+        ans_data_type = std::make_shared<DataTypeString>();
+        break;
+    case tparquet::Type::type::INT96:
+        ans_data_type = std::make_shared<DataTypeInt128>();
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/parquet_column_convert.cpp:54]: Why `tparquet::Type::type::INT96` map to `DataTypeInt128`

File: be/src/vec/exec/format/parquet/parquet_column_convert.h (hunk 1 above)
Comment (AshinGau) [reply be/src/vec/exec/format/parquet/parquet_column_convert.h:265]: Is there `int128` column in parquet, and why convert to timestamp?

File: be/src/vec/exec/format/parquet/parquet_column_convert.h [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+        auto* src_data = static_cast<const ColumnVector<NumberType>*>(src_col)->get_data().data();
+        dst_col->resize(_convert_params->start_idx + rows);
+
+        DecimalScaleParams& scale_params = _convert_params->decimal_scale;
+        auto* data = static_cast<ColumnDecimal<Decimal<DecimalPhysicalType>>*>(dst_col)
+                             ->get_data()
+                             .data();
+
+        for (int i = 0; i < rows; i++) {
+            Int128 value = src_data[i];
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/parquet_column_convert.h:401]: PR https://github.com/apache/doris/pull/25012 has optimized the performance of reading decimal, why still use int128 here?

File: be/src/vec/exec/format/parquet/parquet_column_convert.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        ans_data_type = std::make_shared<DataTypeInt64>();
+        break;
+    case tparquet::Type::type::FLOAT:
+        ans_data_type = std::make_shared<DataTypeFloat32>();
+        break;
+    case tparquet::Type::type::DOUBLE:
+        ans_data_type = std::make_shared<DataTypeFloat64>();
+        break;
+    case tparquet::Type::type::BYTE_ARRAY:
+    case tparquet::Type::type::FIXED_LEN_BYTE_ARRAY:
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/parquet_column_convert.cpp:50]: Column -> UInt8 -> resize(row * typeSize)

File: be/src/vec/exec/format/parquet/parquet_column_convert.h (hunk 1 above)
Comment (AshinGau) [reply be/src/vec/exec/format/parquet/parquet_column_convert.h:265]: int128 -> decimal128 / largeint

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run clickbench

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run p1

**Comment by AshinGau** [comment]
LGTM

**Review by AshinGau — APPROVED** [review]

**Review by morningman — APPROVED** [review]

**Pull request merged**